Automatic Spoilers Season

ZeldaZach opened the pull request:
## Related Ticket(s)
- Fixes (anything?)

## Short roundup of the initial problem
In the past, you had to manually download spoiler files and import them and go through the hassle and such.

## What will change with this Pull Request?
- On cockatrice launch, download a small text file (that we host) to see if we're in spoiler season. If we aren't, just exit. Otherwise...
- Start a timer that is linked to settings. Every X minutes (configurable), we will download the spoilers file from the github page we have and replace the old one. This will also trigger the reloading of the card database and, if any new sets are found, that popup will appear as well.

## Screenshots

<img width="964" alt="screenshot 2017-07-03 21 18 04" src="https://user-images.githubusercontent.com/7460172/27811542-5772bcfa-6035-11e7-998e-87acd84483f7.png">
<img width="429" alt="screenshot 2017-07-03 21 19 03" src="https://user-images.githubusercontent.com/7460172/27811541-57506326-6035-11e7-871e-dfb024266b91.png">

I spent a good few days on this, hopefully it's up to par with what we're looking for!

Psithief commented on cockatrice/CMakeLists.txt:
```diff
@@ -116,6 +116,7 @@ SET(cockatrice_SOURCES
     src/logger.cpp
     src/releasechannel.cpp
     src/userconnection_information.cpp
+    src/SpoilerBackgroundUpdaterThread.cpp
```
This is the only file that's not lowercase in the entire program.

ZeldaZach replied:
Addressed

Psithief commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+SpoilerBackgroundUpdaterThread::SpoilerBackgroundUpdaterThread(QObject *apParent) : QObject(apParent)
+{
+    // Determine if we're in spoilers season or not. If we're not, just kill this thread
+    downloadSpoilerSeasonStatusFile();
```
This really needs to be _after_ the settings check. If someone doesn't want spoilers then there's an unnecessary request being made here.

ZeldaZach replied:
Addressed

Psithief commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+#include "carddatabase.h"
+#include "main.h"
+
+SpoilerBackgroundUpdaterThread::SpoilerBackgroundUpdaterThread(QObject *apParent) : QObject(apParent)
```
Some calls are from a QTimer and some aren't. This naming is confusing.

ZeldaZach replied:
I'm not sure what you want from here...

ZeldaZach replied:
Addressed

Psithief commented:
I have to recommend keeping everything in the main system and only spawning a thread for the actual download.
Just because you use QTimer doesn't mean you need a thread.

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+    }
+
+    // Determine how long to wait between updates (Gets the # of hours from settingsCache / convert Mins -> Millis)
+    float lnTimeToWaitBetween = settingsCache->getDownloadSpoilerTimeMinutes() * 60000;
```
Why float? maybe qint64

ZeldaZach replied:
At one point i was using floats, didn't change back to int. Addressed

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+
+    // How much time is needed until the next update (NOW() - Last Update Time == Time Remaining)
+    long lnLastUpdateTime = settingsCache->getDownloadSpoilerLastUpdateTime();
+    long lnTimeUntilNextUpdate = QDateTime::currentMSecsSinceEpoch() - lnLastUpdateTime;
```
As far as i understand, `lnTimeUntilNextUpdate` should ne named "lnTimeSinceLastUpdate"

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+
+SpoilerBackgroundUpdaterThread::~SpoilerBackgroundUpdaterThread()
+{
+    delete mpTimerForSpoilers;
```
you should take care of deleting `mpNetworkAccessManager ` here, too, if it has been instanciated (!= nullptr).

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+
+    QNetworkReply *lpNetworkReply = mpNetworkAccessManager->get(QNetworkRequest(acUrl));
+
+    connect(lpNetworkReply, SIGNAL(finished()), this, SLOT(actDownloadFinishedTokensFile()));
```
Maybe rename `actDownloadFinishedTokensFile` to `actDownloadFinishedSpoilersFile`

ZeldaZach replied:
Thought I did this, must've stashed it by accident. Addressed

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+    QEventLoop lcSubLoop;
+    connect(lpNetworkReply, SIGNAL(finished()), this, SLOT(actSeeIfSpoilerSeasonIsActive()));
+    connect(lpNetworkReply, SIGNAL(finished()), &lcSubLoop, SLOT(quit()));
+    lcSubLoop.exec();
```
This looks like an hack, and should better be refactored.

ZeldaZach replied:
This 'hack' currently forces the file to be downloaded and parsed _before_ the constructor can continue execution. We need to know if we're in spoiler season or not before continuing, at least that was my idea...

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+{
+    QNetworkReply *lpNetworkReply = static_cast<QNetworkReply *>(sender());
+    QNetworkReply::NetworkError leErrorCode = lpNetworkReply->error();
+    QWidget *lpTempWidget = new QWidget();
```
Is this widget created only to pass a parent to QMessageBox? It should work the same if you pass 0 or nullptr.

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+    if (lcFileToWriteTo.write(mcDownloadedFileData) == -1)
+    {
+        qDebug() << "File write (w) failed for" << asFileName;
+        return false;
```
you should try to close the file even if write failed before returning.

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/settingscache.h:
```diff
@@ -133,12 +142,18 @@ class SettingsCache : public QObject {
     ReleaseChannel * getUpdateReleaseChannel() const { return releaseChannels.at(updateReleaseChannel); }
     QList<ReleaseChannel*> getUpdateReleaseChannels() const { return releaseChannels; }
+    int getDownloadSpoilerTimeIndex() const { return mnDownloadSpoilersTimeIndex; }
+    float getDownloadSpoilerTimeMinutes() const { return msDownloadSpoilersTimeMinutes; }
```
Why do you save in the settings file both the minutes and the entry index in the combobox? You should be able to derive one value from the other.
The problem is that if we ever ned to change the values on the list and keep them ordered, the mapping between indexes and minutes would change and this could cause odd effects when loading old settings in a new client version.

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/window_main.cpp:
```diff
@@ -19,6 +19,7 @@
  ***************************************************************************/
 #include <QMenu>
 #include <QAction>
+#include <iostream>
```
Planning to use `cout`? Please remove if unused

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,233 @@
+        // If the file reads anything else then we're NOT in spoiler season (false)
+        QByteArray laFileData = lpNetworkReply->readAll();
+        mbIsActiveSpoilerSeason = QString(laFileData).contains("enabled", Qt::CaseInsensitive);
+    }
```
To avoid the hack code at lines 135-140, you could start the timer here instead of in the constructor.

ZeldaZach replied:
Moved, not sure if it'll do much though

ZeldaZach commented:
I have pushed a change that addresses all the comments brought up.
- The program will NOT check for a spoilers byte file on github if you don't have the feature enabled. 
- If you choose to enable it while the program is running, it will auto-download at that moment and then restart the timer from that point.
- Fixed up file name
- More stuff (see the changelog).

Please provide additional feedback at this time :)

ZeldaZach commented on cockatrice/src/dlg_settings.cpp:
```diff
@@ -464,25 +464,154 @@ void UserInterfaceSettingsPage::retranslateUi()
+
+void DeckEditorSettingsPage::setDownloadSpoilerTime(QString asValue)
+{
+    // Sets the downloadspoilerstimeHours field (stored in HOURS (float))
```
haha... minutes, int

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,249 @@
+#include <QDateTime>
```
This file needs to be renamed in order to be all lowercase, or compilation will fail on case sensitive filesystems (eg. Linux)

ZeldaZach replied:
I swear I did... I guess github doesn't recognize case changes

ctrlaltca replied:
If you are working on a Mac, the filesystem is case insensitive, so it won't pick up changes to the case of file names.. you can probably workaround it by renaming the file to something else, committing, then rename again to the correct name and re-commit.

ctrlaltca commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,249 @@
+    if (leErrorCode != QNetworkReply::NoError)
+    {
+        QMessageBox::critical(nullptr, tr("Error"), tr("Network error: %1.").arg(lpNetworkReply->errorString()));
+    }
```
Just a personal preference, take it as a "soft" suggestion:
In methods like this where you have a lot of if/then/else branches, try to keep the logic flow as simple as possible. For example, here adding a 
```
lpNetworkReply->deleteLater();
return;
```
you can remove the `else` branch and de-indent everything of one step.
The same can be done at line 185.
So instead of having three levels of nested `if`s, you can get an easier to follow flow like this:
```
if(!condition1)
  return;
if(!condition2)
  return;
get_the_job_done();
```

ZeldaZach replied:
Dealt with

ctrlaltca commented on cockatrice/src/dlg_settings.cpp:
```diff
@@ -464,25 +464,154 @@ void UserInterfaceSettingsPage::retranslateUi()
+    if (abIsEnabled)
+    {
+        long lnTimeUntilNextUpdate = QDateTime::currentMSecsSinceEpoch() - settingsCache->getDownloadSpoilerLastUpdateTime();
+        float lnTimeToWaitBetween = settingsCache->getDownloadSpoilerTimeMinutes() * 60000;
```
Why float? Use long, long long (or, better, qint64) and avoid the static_cast<long> in the following lines

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/dlg_settings.cpp:
```diff
@@ -464,25 +464,154 @@ void UserInterfaceSettingsPage::retranslateUi()
+    // On a change to the check box, hide/unhide the other fields
+    // On a change to the combo box, send a signal to the timer threads
+    connect(&mcDownloadSpoilersCheckBox, SIGNAL(stateChanged(int)), settingsCache, SLOT(setDownloadSpoilerStatus(int)));
+    connect(&mcDownloadSpoilersCheckBox, SIGNAL(stateChanged(int)), this, SLOT(setSpoilersEnabled(int)));
```
You may want to bind the `toggled(bool)` signal instead to avoid the `static_cast<bool>` in the receiving code
(Also, the int returned by stateChanged has more values than just 0-unchecked 1-checked)

ZeldaZach replied:
Addressed

ZeldaZach commented on cockatrice/src/SpoilerBackgroundUpdaterThread.cpp:
```diff
@@ -0,0 +1,249 @@
+    connect(settingsCache, SIGNAL(downloadSpoilerStatusChanged()), this, SLOT(changeActiveStatus()));
+
+    // When timer ends, cause download of Spoilers XML file
+    connect(mpTimerForSpoilers, SIGNAL(timeout()), this, SLOT(timeoutOccurredTimeToDownloadSpoilers()));
```
Addressed

ZeldaZach commented:
I have made all the requested changes.

ctrlaltca commented on cockatrice/src/settingscache.h:
```diff
@@ -4,6 +4,7 @@
 #include <QObject>
 #include <QSize>
 #include <QStringList>
+#include <QTimer>
```
Probably unneeded

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/spoilerbackgroundupdater.cpp:
```diff
@@ -0,0 +1,241 @@
+
+SpoilerBackgroundUpdater::~SpoilerBackgroundUpdater()
+{
+    delete mpTimerForSpoilers;
```
If the user disabled spoiler checking, mpTimerForSpoilers  will never be created, and trying to delete it will probably cause a crash.
Anyway, i saw that when you instanciate mpTimerForSpoilers and mpNetworkAccessManager you always pass "this" as the object parent, so you can just skip deleting these 2 objects since Qt will take care of deleting them when their parent gets deleted.
Sorry, i know i first asked you to explicitly delete mpNetworkAccessManager, i overlooked the code.

ZeldaZach replied:
Addressed

ctrlaltca commented on cockatrice/src/spoilerbackgroundupdater.cpp:
```diff
@@ -0,0 +1,241 @@
+        if (mbIsActiveSpoilerSeason)
+        {
+            // Start the timer for spoilers downloads
+            mpTimerForSpoilers = new QTimer(this);
```
runTimer() already takes care of creating the timer if it doesn't exists yet, no need to do it here
(Also, this was not checking if the timer already existed, so it could create a duplicate)

ZeldaZach replied:
Addressed

ZeldaZach commented:
When I try to click the checkbox on windows:
```
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	cockatrice.exe
  Application Version:	0.0.0.0
  Application Timestamp:	5960b087
  Fault Module Name:	Qt5Network.dll
  Fault Module Version:	5.6.2.0
  Fault Module Timestamp:	57e77689
  Exception Code:	c0000005
  Exception Offset:	00000000000252e6
  OS Version:	6.1.7601.2.1.0.256.48
  Locale ID:	1033
  Additional Information 1:	8355
  Additional Information 2:	83551fe3c4bc47bb701562fbf1154230
  Additional Information 3:	0e93
  Additional Information 4:	0e9318370a9d9370baef69ec5f67ab73

Read our privacy statement online:
  http://go.microsoft.com/fwlink/?linkid=104288&clcid=0x0409

If the online privacy statement is not available, please read our privacy statement offline:
  C:\Windows\system32\en-US\erofflps.txt

```

ZeldaZach commented:
@ctrlaltca Some debugging stuff:
<img width="739" alt="screenshot 2017-07-08 12 43 26" src="https://user-images.githubusercontent.com/7460172/27987404-3415be8e-63db-11e7-81c8-e6df9d19a451.png">
<img width="565" alt="screenshot 2017-07-08 12 43 43" src="https://user-images.githubusercontent.com/7460172/27987405-3415d3b0-63db-11e7-98f6-97b9b69d3258.png">

Brings me to https://github.com/Cockatrice/Cockatrice/pull/2796/files#diff-2c069c6d2a8e071f48b42363bb52741aR137 in the code base

ctrlaltca commented:
![schermata 2017-07-09 alle 10 22 21](https://user-images.githubusercontent.com/1631111/27992280-906e70b2-6490-11e7-89bb-00d631828efb.png)

As you can see in the bottom left panel, both `mpNetworkAccessManager` and  `lpNetworkReply` have a weird value of `0xcccccccc`. At this point in the code you would expect `mpNetworkAccessManager` to have been instanciated a few lines before, while `lpNetworkReply` should be an invalid pointer, since we're just creating it.
Instead, they both contain an invalid value; this is due to the fact that `mpNetworkAccessManager` is not initialized to `nullptr` when the class is instanciated, so it will contain random garbage from the memory.

ctrlaltca commented on cockatrice/src/spoilerbackgroundupdater.h:
```diff
@@ -0,0 +1,41 @@
+
+private:
+    QTimer *mpTimerForSpoilers;
+    QNetworkAccessManager *mpNetworkAccessManager;
```
When you add a pointer member to a class, always ensure you initialize it in the class constructor, either instnciating the object or setting it to nullptr

Daenyth commented:
What happens with the spoiler file when season is over? Is there anything that prevents double entries, when you have a spoiler file + oracle run after the sets are out?

I'm concerned about user education. Oracle is already hard to understand for users. This would make it seem like they get the new cards automatically but then when season is over, they need to run oracle manually. Thoughts? I don't think that's a blocker for adding this - I think it's an awesome feature addition. I'm just wondering how we can best educate users.

I haven't looked at the code yet, but one thing I'd like to make sure is that all the steps around this have good debug log output so we can help people troubleshoot.

ZeldaZach commented:
@ctrlaltca You still working on a sub-PR for this?
